$OpenBSD: patch-filter_hpgl-attr_c,v 1.1.2.1 2009/06/28 03:53:16 william Exp $

CVE-2008-3641, without SP_select_pen() change in STR #2911,
and fix an off-by-one (STR #2966)

--- filter/hpgl-attr.c.orig	Thu Feb 17 21:18:11 2005
+++ filter/hpgl-attr.c	Sat Jun 20 23:13:13 2009
@@ -205,13 +205,23 @@ NP_number_pens(int     num_params,	/* I - Number of pa
 
   if (num_params == 0)
     PenCount = 8;
-  else if (num_params == 1 && params[0].value.number <= 1024)
-    PenCount = (int)params[0].value.number;
+  else if (num_params == 1)
+  {
+    if (params[0].value.number < 1 || params[0].value.number > MAX_PENS)
+    {
+      fprintf(stderr,
+	      "DEBUG: HP-GL/2 \'NP\' command with invalid number of "
+	      "pens (%d)!\n", (int)params[0].value.number);
+      PenCount = 8;
+    }
+    else
+      PenCount = (int)params[0].value.number;
+  }
   else
     fprintf(stderr, "WARNING: HP-GL/2 \'NP\' command with invalid number of parameters (%d)!\n",
             num_params);
 
-  for (i = 0; i <= PenCount; i ++)
+  for (i = 0; i < PenCount; i ++)
     Pens[i].width = PenWidth;
 
   PC_pen_color(0, NULL);
@@ -242,7 +252,7 @@ PC_pen_color(int     num_params,	/* I - Number of para
 
   if (num_params == 0)
   {
-    for (i = 0; i <= PenCount; i ++)
+    for (i = 0; i < PenCount; i ++)
       if (i < 8)
       {
         Pens[i].rgb[0] = standard_colors[i][0];
@@ -263,8 +273,15 @@ PC_pen_color(int     num_params,	/* I - Number of para
   }
   else if (num_params == 1 || num_params == 4)
   {
-    i = (int)params[0].value.number;
+    i = (int)params[0].value.number - 1;
 
+    if (i < 0 || i >= PenCount)
+    {
+      fprintf(stderr,
+              "DEBUG: HP-GL/2 \'PC\' command with invalid pen (%d)!\n", i + 1);
+      return;
+    }
+
     if (num_params == 1)
     {
       Pens[i].rgb[0] = standard_colors[i & 7][0];
@@ -336,8 +353,16 @@ PW_pen_width(int     num_params,	/* I - Number of para
 
   if (num_params == 2)
   {
-    pen = (int)params[1].value.number;
+    pen = (int)params[1].value.number - 1;
 
+    if (pen < 0 || pen >= PenCount)
+    {
+      fprintf(stderr,
+              "DEBUG: HP-GL/2 \'PW\' command with invalid pen (%d)!\n",
+	      pen + 1);
+      return;
+    }
+
     Pens[pen].width = w;
 
     if (PageDirty && pen == PenNumber)
@@ -351,7 +376,7 @@ PW_pen_width(int     num_params,	/* I - Number of para
     * Set width for all pens...
     */
 
-    for (pen = 0; pen <= PenCount; pen ++)
+    for (pen = 0; pen < PenCount; pen ++)
       Pens[pen].width = w;
 
     if (PageDirty)
@@ -404,12 +429,15 @@ SP_select_pen(int     num_params,	/* I - Number of par
               param_t *params)		/* I - Parameters */
 {
   if (num_params == 0)
-    PenNumber = 1;
-  else if (params[0].value.number <= PenCount)
-    PenNumber = (int)params[0].value.number;
+    PenNumber = 0;
+  else if (num_params > 1)
+    fprintf(stderr, "WARNING: HP-GL/2 \'SP\' command with invalid number of parameters (%d)!\n",
+            num_params);
+  else if (params[0].value.number <= 0 || params[0].value.number >= PenCount)
+    fprintf(stderr, "DEBUG: HP-GL/2 \'SP\' command with invalid pen (%d)!\n",
+            (int)params[0].value.number);
   else
-    fprintf(stderr, "WARNING: HP-GL/2 \'SP\' command with invalid number or value of parameters (%d, %d)!\n",
-            num_params, (int)params[0].value.number);
+    PenNumber = (int)params[0].value.number;
 
   if (PageDirty)
     printf("%.3f %.3f %.3f %.2f SP\n", Pens[PenNumber].rgb[0],
